01_fetch/src/fetch_noaa_nerrs.py

```diff
@@ -0,0 +1,51 @@
+import pandas as pd
+import os 
+import numpy as np
+import urllib
+import requests, json
+
+def fetch_metadata(station_id, metadata_outfile):
+    '''fetch metadata for NOAA NERRS station'''
+    metadata_url = f'https://api.tidesandcurrents.noaa.gov/mdapi/prod/webapi/stations/{station_id}/.json'
+    response = requests.get(metadata_url)
+    text = response.text
+    data = json.loads(text)
+    nested_list = pd.json_normalize(data, record_path=['stations'])
+    df = nested_list.stack().reset_index()
+    metadata = df.iloc[1:,1:].rename(columns={df.columns[1]:"NOAA NERRS metadata:",
+                                              df.columns[2]:(station_id)}).replace(".self","", regex=True)
+    metadata.to_csv(metadata_outfile, index=False)
+
+def fetch_noaa_nerrs_data(start_dt, end_dt, datum, station_id, time_zone, product, units, file_format, data_outfile):
+    '''fetch NOAA NERRS data from select station. Change product argument for NOAA data product. (ie. product = current for current data)'''
+    dates = pd.date_range(start=start_dt, end=end_dt, freq='D')
+    dates = dates.strftime('%Y%m%d').astype(int)
+    url = "https://api.tidesandcurrents.noaa.gov/api/prod/datagetter?product=predictions&application=NOS.COOPS.TAC.WL"
+    data_url = f'{url}&begin_date={dates[0]}&end_date={dates[-1]}&datum={datum}&station={station_id}&product={product}&time_zone={time_zone}&units={units}&interval=&format={file_format}'
+    outfile_format = data_outfile.format(start_dt=start_dt, end_dt=end_dt, station_id=station_id, data_outfile=data_outfile, product=product)
+    urllib.request.urlretrieve(data_url, outfile_format)
```

Comment:
Could we just input the dates in the format needed so that you don't have to do these transformations? I think you were trying to match my input date format, but I think it might be easier here to just use the needed format and specify that format in a comment.

Also, we can actually probably just format this url outside of the function so you only have to pass in the url and not all the parameters. I know I passed unformatted urls into my functions but I think it was because I was doing more transformations inside the functions.
```suggestion
def fetch_noaa_nerrs_data(start_dt, end_dt, datum, station_id, time_zone, product, units, file_format, data_outfile):
    '''fetch NOAA NERRS data from select station. Change product argument for NOAA data product. (ie. product = current for current data)'''
        data_url = f'"https://api.tidesandcurrents.noaa.gov/api/prod/datagetter?product=predictions&application=NOS.COOPS.TAC.WL&begin_date={start_dt}&end_date={end_dt}&datum={datum}&station={station_id}&product={product}&time_zone={time_zone}&units={units}&interval=&format={file_format}"
    urllib.request.urlretrieve(data_url, data_outfile)
```

Reply:
Yea, that makes sense. It seemed unlikely that a user would type the start date and end date in that format (%Y%m%d), so I tried to make it easy.